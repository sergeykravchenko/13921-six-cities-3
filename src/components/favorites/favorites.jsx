import React, {PureComponent} from 'react';
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {getFavoritesByCity} from "../../reducer/state/selectors";
import {ActionCreator} from "../../reducer/state/state";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import Header from "../header/header.jsx";
import FavoritesEmpty from "../favorites-empty/favorites-empty.jsx";
import PlaceCard from '../place-card/place-card.jsx';
import Footer from "../footer/footer.jsx";

class Favorites extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.onMount();
  }

  render() {
    const {favorites, handlePlaceTitleClick} = this.props;

    if (!favorites || favorites.length <= 0) {
      return (
        <div className="page page--favorites-empty">
          <Header/>
          <FavoritesEmpty />
          <Footer />
        </div>
      );
    }
    return (
      <div className="page page__main--favorites">
        <Header/>
        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">
                {favorites.map((item) =>
                  Object.entries(item).map(([city, offers]) =>
                    <li key={city} className="favorites__locations-items">
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link" href="#">
                            <span>{city}</span>
                          </a>
                        </div>
                      </div>
                      <div className="favorites__places">
                        {offers.map((offer) =>
                          <PlaceCard
                            bemblock={`favorites`}
                            key={offer.id}
                            place={offer}
                            handlePlaceTitleClick={handlePlaceTitleClick}
                          />
                        )}
                      </div>
                    </li>
                  ))}
              </ul>
            </section>
          </div>
        </main>
        <Footer />
      </div>
    );
  }
}

Favorites.propTypes = {
  favorites: PropTypes.array,
  handlePlaceTitleClick: PropTypes.func,
  onMount: PropTypes.func,
};

const mapStateToProps = (state) => {
  return {
    favorites: getFavoritesByCity(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  onMount() {
    dispatch(DataOperation.loadFavorites());
  },
  handlePlaceTitleClick(offer) {
    dispatch(ActionCreator.getActiveOffer(offer));
    dispatch(DataOperation.loadComments(offer.id));
    dispatch(DataOperation.loadNearByOffer(offer.id));
  },
});

export {Favorites};
export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
