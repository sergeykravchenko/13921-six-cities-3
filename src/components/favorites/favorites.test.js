import React from "react";
import renderer from "react-test-renderer";
import {Router} from "react-router-dom";
import history from "../../history.js";
import {Favorites} from "./favorites.jsx";

const favorites = [
  {
    'Hamburg': [{
      "id": 1,
      "name": `Canal View Prinsengracht`,
      "picture": `img/room.jpg`,
      "price": 120,
      "priceText": `night`,
      "rating": 4,
      "zoom": 16,
      "features": {
        "type": `Private room`,
        "bedrooms": 8,
        "maxGuests": 7,
      },
      "isPremium": true,
      "isInBookmark": true,
      "houseHolds": [
        `Wi-Fi`,
        `Washing machine`,
      ],
      "gallery": [
        `img/apartment-01.jpg`,
        `img/apartment-02.jpg`,
      ],
      "host": {
        "id": 2,
        "name": `Max`,
        "img": `img/avatar-max.jpg`,
        "isPro": true,
      },
      "description": `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      "coords": [23, 12],
    }, {
      "id": 2,
      "name": `Canal View Prinsengracht`,
      "picture": `img/room.jpg`,
      "price": 120,
      "priceText": `night`,
      "rating": 4,
      "zoom": 16,
      "features": {
        "type": `Private room`,
        "bedrooms": 8,
        "maxGuests": 7,
      },
      "isPremium": true,
      "isInBookmark": true,
      "houseHolds": [
        `Wi-Fi`,
        `Washing machine`,
      ],
      "gallery": [
        `img/apartment-01.jpg`,
        `img/apartment-02.jpg`,
      ],
      "host": {
        "id": 2,
        "name": `Max`,
        "img": `img/avatar-max.jpg`,
        "isPro": true,
      },
      "description": `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      "coords": [25, 13],
    }]
  }
];

it(`Render Favorites`, () => {
  const tree = renderer
    .create(
        <Router
          history={history}
        >
          <Favorites
            favorites={favorites}
            onMount={() => {}}
            handlePlaceTitleClick={() => {}}
          />
        </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
