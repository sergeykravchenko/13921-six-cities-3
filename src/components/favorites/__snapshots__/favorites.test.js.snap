// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Favorites 1`] = `
<div
  className="page page--favorites-empty"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                  style={
                    Object {
                      "backgroundImage": "url(https://htmlacademy-react-3.appspot.com/six-citiesimg/1.png)",
                      "borderRadius": "50%",
                    }
                  }
                />
                <span
                  className="header__login"
                >
                  Oliver.conner@gmail.com
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites page__main--favorites-empty"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites favorites--empty"
      >
        <h1
          className="visually-hidden"
        >
          Favorites (empty)
        </h1>
        <div
          className="favorites__status-wrapper"
        >
          <b
            className="favorites__status"
          >
            Nothing yet saved.
          </b>
          <p
            className="favorites__status-description"
          >
            Save properties to narrow down search or plan yor future trips.
          </p>
        </div>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
